apps/sms/test/marionette/mocks/mock_navigator_moz_mobile_message.js

```diff
@@ -27,215 +24,210 @@ Services.obs.addObserver(function(document) {
-          };
-
-          var o = new window.Object();
-          o.message = createMessage(parameters, options);
-
-          callEventHandlers('sending', o);
-        }, delayMs);
+      var requests = recipients.map((receiver) => {
+        var request = Services.DOMRequest.createRequest(window);
+
+        var parameters = {
+          type: 'sms',
+          body: content,
+          receiver: receiver
+        };
+
+        createMessage(parameters, options).then(function(message) {
+          Services.DOMRequest.fireSuccessAsync(request, null);
+
+          callEventHandlers(
+            'sending', Cu.cloneInto({ message: message }, window)
+          );
+        });
       });
 
-      return requests;
+      return Cu.waiveXrays(requests);
     },
 
     getThreads: function() {
-      var cursor = Cu.waiveXrays(Cu.cloneInto({
-        onsuccess: null,
-        onerror: null
-      }, window));
+      var threads = null;
+
+      var handleCursor = function() {
+        var iteratorResult = threads.next();
 
-      window.setTimeout(function() {
-        if (typeof cursor.onsuccess === 'function') {
-          cursor.onsuccess.call(cursor);
+        if (!iteratorResult.done) {
+          Services.DOMRequest.fireSuccessAsync(cursor, iteratorResult.value);
+        } else {
+          Services.DOMRequest.fireDone(cursor);
         }
-      }, delayMs);
+      };
+
+      var cursor = Services.DOMRequest.createCursor(window, handleCursor);
+
+      getStorage().then(function(storage) {
+        threads = storage.threads.values();
+
+        handleCursor();
+      });
 
       return cursor;
     },
 
     getMessages: function(filter) {
-      var cursor = Cu.waiveXrays(Cu.cloneInto({
-        onsuccess: null,
-        onerror: null,
-        result: null,
-        continue: null,
-        done: false
-      }, window));
+      var messages = null;
 
-      window.setTimeout(function() {
-        if (typeof cursor.onsuccess === 'function') {
-          var thread = threads.get(filter.threadId),
-              currentIndex = -1;
-
-          cursor.continue = function() {
-            if (!thread || ++currentIndex >= thread.messages.length) {
-              cursor.done = true;
-              cursor.continue = null;
-              cursor.result = null;
-            } else {
-              cursor.result = thread.messages[currentIndex];
-            }
-            cursor.onsuccess.call(cursor);
-          };
+      var handleCursor = function() {
+        var iteratorResult = messages.next();
 
-          cursor.continue();
+        if (!iteratorResult.done) {
+          Services.DOMRequest.fireSuccessAsync(cursor, iteratorResult.value);
+        } else {
+          Services.DOMRequest.fireDone(cursor);
         }
-      }, delayMs);
+      };
+
+      var cursor = Services.DOMRequest.createCursor(window, handleCursor);
+
+      getStorage().then(function(storage) {
+        var thread = storage.threads.get(filter.threadId);
```

Comment:
maybe throw early if we don't have this filter? This would help finding issues in the test if we change the client code (for example if we ever paginate using the timestamps using the API -- I'm not sure we'll ever do that this way but..).